[testing] Reduce flaky tests by retrying git failures

### Housekeeping

- [X] I am a maintainer of dbt-core

### Short description

We have a lot of tests that are failing because of Git connection issues.  Sometimes tox fails to install all dependencies and that causes the entire test run to fail without actually running any tests.  This makes our monitoring noisy.

Suggested approach: leveraging something the `nick-fields/retry@v3` action ([example](https://github.com/dbt-labs/actions/blob/7738a62a39f411ddf53aa1fa104783f9b9eb007e/.github/workflows/release-branch-tests.yml#L135-L143) but in the tox invocation [here](https://github.com/dbt-labs/dbt-core/blob/fe28d9e115f9a31cedcc641f67fcd4d229ad0027/.github/workflows/main.yml#L204-L207))

### Acceptance criteria

Anytime we use git when testing, have retry logic

### Suggested Tests

This task is specifically for tests

-- can force a test to fail in a commit & observe the retry works as expected at the integration group level

### Impact to Other Teams

Adapters team won't be impacted but may be interested if we come up with a solution

### Will backports be required?

backport as far as we can to reduce this noise

### Context

log output from test failing on tox
```
Run tox -- --ddtrace
integration: install_deps> python -I -m pip install -r dev-requirements.txt -r editable-requirements.txt
  Running command git clone --filter=blob:none --quiet https://github.com/dbt-labs/dbt-adapters.git /tmp/pip-req-build-g9zkv3vu
  error: RPC failed; curl 16 Error in the HTTP2 framing layer
  fatal: expected 'packfile'
  fatal: could not fetch 22b2ad3f683cca452f28320c0aba8bb95933ca6e from promisor remote
Collecting git+https://github.com/dbt-labs/dbt-adapters.git@main (from -r dev-requirements.txt (line 1))
  Cloning https://github.com/dbt-labs/dbt-adapters.git (to revision main) to /tmp/pip-req-build-g9zkv3vu
integration: exit 1 (2.55 seconds) /home/runner/work/dbt-core/dbt-core> python -I -m pip install -r dev-requirements.txt -r editable-requirements.txt pid=1980
  warning: Clone succeeded, but checkout failed.
  You can inspect what was checked out with 'git status'
  and retry with 'git restore --source=HEAD :/'

  error: subprocess-exited-with-error
  
  × git clone --filter=blob:none --quiet https://github.com/dbt-labs/dbt-adapters.git /tmp/pip-req-build-g9zkv3vu did not run successfully.
  │ exit code: 1[28](https://github.com/dbt-labs/dbt-core/actions/runs/8633734890/job/23667503237#step:8:29)
  ╰─> See above for output.
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
error: subprocess-exited-with-error

× git clone --filter=blob:none --quiet https://github.com/dbt-labs/dbt-adapters.git /tmp/pip-req-build-g9zkv3vu did not run successfully.
│ exit code: 128
╰─> See above for output.

note: This error originates from a subprocess, and is likely not a problem with pip.
  integration: FAIL code 1 (5.44 seconds)
  evaluation failed :( (5.50 seconds)

Error: Process completed with exit code 1.
```

Sample of tests marked as flaky but are likely just connection issues.  There may not be a solution when there's a longer GitHub outage.  Look through #7808 for other possible failures.

#9906 

max retries exceeded
#9905 
#9903 

timeout
#9902 
#9900 

Note: integration tests are run with the workflow_dispatch trigger in scheduled testing [here](https://github.com/dbt-labs/actions/blob/7738a62a39f411ddf53aa1fa104783f9b9eb007e/.github/workflows/release-branch-tests.yml#L114C11-L114C63).  typically it would be run with workflow_call trigger but isn't because it's special ([comment](https://github.com/dbt-labs/actions/blob/7738a62a39f411ddf53aa1fa104783f9b9eb007e/.github/workflows/release-branch-tests.yml#L107-L110))

--- a/.github/workflows/main.yml
+++ b/.github/workflows/main.yml
@@ -68,8 +68,12 @@ jobs:
           python -m pip --version
           tox --version
 
-      - name: Run tox
-        run: tox
+      - name: Run integration tests
+        uses: nick-fields/retry@v3
+        with:
+          timeout_minutes: 30
+          max_attempts: 3
+          command: tox
 
   unit:
     name: unit test / python ${{ matrix.python-version }}

--- a/.github/workflows/main.yml
+++ b/.github/workflows/main.yml
@@ -103,8 +103,12 @@ jobs:
           python -m pip --version
           tox --version
 
-      - name: Run tox
-        run: tox
+      - name: Run unit tests
+        uses: nick-fields/retry@v3
+        with:
+          timeout_minutes: 10
+          max_attempts: 3
+          command: tox -e unit
 
       - name: Get current date
         if: always()

--- a/.github/workflows/structured-logging-schema-check.yml
+++ b/.github/workflows/structured-logging-schema-check.yml
@@ -1,6 +1,6 @@
 # This Action checks makes a dbt run to sample json structured logs
 # and checks that they conform to the currently documented schema.
-# 
+#
 # If this action fails it either means we have unintentionally deviated
 # from our documented structured logging schema, or we need to bump the
 # version of our structured logging and add new documentation to
@@ -61,7 +61,11 @@ jobs:
       # integration tests generate a ton of logs in different files. the next step will find them all.
       # we actually care if these pass, because the normal test run doesn't usually include many json log outputs
       - name: Run integration tests
-        run: tox -e py38-postgres -- -nauto
+        uses: nick-fields/retry@v3
+        with:
+          timeout_minutes: 30
+          max_attempts: 3
+          command: tox -e py38-postgres -- -nauto
 
       # apply our schema tests to every log event from the previous step
       # skips any output that isn't valid json

